src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,117 @@
 package core.basesyntax;
 
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
+    private static final int START_CAPACITY = 10;
+
+    private Object[] values;
+    private int size;
+
+    public ArrayList() {
+        this.values = new Object[START_CAPACITY];
+    }
+
+    private Object[] grow() {
+        int oldCapacity = values.length;
+        Object[] buffer = values;
+        values = new Object[oldCapacity + (oldCapacity >> 1)];
+        System.arraycopy(buffer, 0, values, 0, size);
+        return values;
+    }
+
     @Override
     public void add(T value) {
-
+        if (values.length == size) {
+            values = grow();
+        }
+        values[size] = value;
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        if (index < 0 || index > size()) {
+            throw new ArrayListIndexOutOfBoundsException(
+                    "Can't find such index " + index + " in this array");
+        }
 
+        if (values.length == size) {
+            values = grow();
+        }
+        System.arraycopy(values, index,
+                values, index + 1,
+                size - index);
+        values[index] = value;
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        for (int i = 0; i < list.size(); i++) {
+            values[size] = list.get(i);
+            size++;
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        if (index < 0 || index >= size) {
+            throw new ArrayListIndexOutOfBoundsException(
+                    "Can't find such index " + index + " in this array");
+        }
+        return (T) values[index];
     }
 
     @Override
     public void set(T value, int index) {
+        if (index < 0 || index >= size) {
+            throw new ArrayListIndexOutOfBoundsException(
+                    "Can't find such index " + index + " in this array");
+        }
+        values[index] = value;
+    }
 
+    private void fastRemove(Object[] helpToRemoveArray, int i) {
+        final int newSize;
+        if ((newSize = size - 1) > i) {
+            System.arraycopy(helpToRemoveArray, i + 1, helpToRemoveArray, i, newSize - i);
+        }
+        helpToRemoveArray[size = newSize] = null;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        if (index < 0 || index >= size) {
+            throw new ArrayListIndexOutOfBoundsException(
+                    "Can't find such index " + index + " in this array");
+        }
```

Comment:
Don’t create repeating code. If the logic of your code repeats - move it to the separate private method.